crates/codegen/syntax/src/nodes/kinds/mod.rs

```diff
@@ -17,95 +25,191 @@ impl TemplateContext for NodeKindsTemplate {
 
-        // Insert "SKIPPED" as the first token kind (after sorting):
-        enum_variants.insert(0, "SKIPPED".to_owned());
+        variants.insert(
+            0, // Insert "SKIPPED" as the first token kind (after sorting)
+            NodeKindsVariant {
+                name: "SKIPPED".to_owned(),
+                documentation:
+                    "Used to hold parts of input that cannot be parsed (incomplete, or erroneous)."
+                        .to_owned(),
+            },
+        );
 
         return Self {
-            enum_name: "TokenKind".to_string(),
-            enum_variants,
+            type_name: "TokenKind".to_string(),
+            variants,
         };
     }
 }
+
+fn collect_top_level(
+    language: &LanguageDefinitionRef,
+    production: &ProductionRef,
+    variants: &mut Vec<NodeKindsVariant>,
+) {
+    if production.inlined {
+        return;
+    }
+
+    let documentation = extract_documentation(language, production);
+
+    variants.push(NodeKindsVariant {
+        name: production.name.to_owned(),
+        documentation: documentation[&production.name].to_owned(),
+    });
+}
+
+fn collect_recursive(
+    language: &LanguageDefinitionRef,
+    production: &ProductionRef,
+    variants: &mut Vec<NodeKindsVariant>,
+) {
+    let documentation = extract_documentation(language, production);
+
+    if !production.inlined {
+        variants.push(NodeKindsVariant {
+            name: production.name.to_owned(),
+            documentation: documentation[&production.name].to_owned(),
+        });
+    }
+
+    let version_map = match &production.definition {
+        ProductionDefinition::PrecedenceParser { version_map, .. } => version_map,
+        ProductionDefinition::Parser { .. }
+        | ProductionDefinition::Scanner { .. }
+        | ProductionDefinition::TriviaParser { .. } => {
+            // Nothing recursive to collect
+            return;
+        }
+    };
+
+    let expressions: Vec<_> = match &version_map {
+        codegen_schema::types::VersionMap::Unversioned(unversioned) => unversioned
+            .operator_expressions
+            .iter()
+            .map(|expression| &expression.name)
+            .unique()
+            .collect(),
+        codegen_schema::types::VersionMap::Versioned(versioned) => versioned
+            .values()
+            .filter_map(|version| version.as_ref())
+            .flat_map(|precedence_parser| &precedence_parser.operator_expressions)
+            .map(|expression| &expression.name)
+            .unique()
+            .collect(),
+    };
+
+    for expression in expressions {
+        variants.push(NodeKindsVariant {
+            name: expression.to_owned(),
+            documentation: documentation[expression].to_owned(),
+        });
+    }
+}
+
+// Constructs all documentation headers for the given production, including any sub-statements
+// It returns a map from each statement name to its markdown documentation:
+fn extract_documentation(
```

Comment:
This looks like a good candidate to refactor the shared EBNF serializer to produce documentation in a single place for all three consumers. However, we are changing how we do versioning soon, so I decided to keep this for now.